File: Depth_First_Search/695.Max_Area_of_Island/C++/695.Max_Area_of_Island.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>> & grid) 
    {
        //============================================//
        // Input type:                                //
        //   - vector<vector<int>> ref. (grid vector) //
        // Return type:                               //
        //   - int (result maximun area)              //
        //============================================//

        /*Initialize*/
        ///// Length of rows, columns /////
        int len_rows = grid.size(), len_cols = (grid[0]).size();

        ///// Record area /////
        int record_area = 0;

        ///// Result maximun area /////
        int res_max_area = 0;


        /*Recursion-based DFS loop traversal*/
        for (int rows_idx = 0; (rows_idx < len_rows); (rows_idx++)) //Rows
        {
            for (int cols_idx = 0; (cols_idx < len_cols); (cols_idx++)) //Columns
            {
                ///// Check if the current indexed-value matched conditions or not /////
                if ((grid[rows_idx])[cols_idx] == 1)
                {
                    dfsTraversal(grid, len_rows, len_cols, rows_idx, cols_idx, record_area); //Recursion function call

                    res_max_area = max(res_max_area, record_area); //Keep updating, overwriting

                    record_area &= 0; //Keep updating, reset
                }
                else { ; }

            } //Columns

        } //Rows

        return res_max_area;
    }


    void dfsTraversal(vector<vector<int>> & grid, int & len_rows, int & len_cols, int rows_idx, int cols_idx, int & record_area)
    {
        //============================================//
        // Input type:                                //
        //   - vector<vector<int>> ref. (grid vector) //
        //   - int ref. (length of rows)              //
        //   - int ref. (length of columns)           //
        //   - int (index of rows)                    //
        //   - int (index of columns)                 //
        //   - int ref. (record area)                 //             
        // Return type:                               //
        //   - void (no return)                       //
        //============================================//
        if (((rows_idx < 0) || (rows_idx >= len_rows)) || ((cols_idx < 0) || (cols_idx >= len_cols))) { return; } //Issue, Boundary-case handling
        if ((grid[rows_idx])[cols_idx] != 1) { return; } //Issue, Boundary-case handling
 
        (record_area++); //Update, Accumulate

        (grid[rows_idx])[cols_idx] = (-1); //Update, Overwrite

        dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx - 1), record_area); //Recursion function call (i.e. left)
        dfsTraversal(grid, len_rows, len_cols, (rows_idx - 1), cols_idx, record_area); //Recursion function call (i.e. up)
        dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx + 1), record_area); //Recursion function call (i.e. right)
        dfsTraversal(grid, len_rows, len_cols, (rows_idx + 1), cols_idx, record_area); //Recursion function call (i.e. down)
    }
};

File: Depth_First_Search/695.Max_Area_of_Island/Python/695.Max_Area_of_Island.py
class Solution(object):
    def maxAreaOfIsland(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        #======================================#
        # Recursion-based DFS traversal method #
        #======================================#

        ############
        #Initialize
        ##### Length of rows, columns #####
        len_rows, len_cols = len(grid), len(grid[0])

        ##### Record area #####
        record_area = [0]

        ##### Result maximun area #####
        res_max_area = 0


        #####################################
        #Recursion-based DFS loop traversal
        for rows_idx in range(len_rows):
            for cols_idx in range(len_cols):

                ##### Check if the current indexed-value matched conditions or not #####
                if ((grid[rows_idx])[cols_idx] == 1):
                    self.dfsTraversal(grid, len_rows, len_cols, rows_idx, cols_idx, record_area) #Recursion function call

                    res_max_area = max(res_max_area, record_area[0]) #Keep updating/overwriting

                    record_area[0] &= 0 #Keep updating/reset
                else:
                    pass

        return res_max_area


    def dfsTraversal(self, grid, len_rows, len_cols, rows_idx, cols_idx, record_area):
        """
        :type grid: List[List[int]]
        :type len_rows: int
        :type len_cols: int
        :type rows_idx: int
        :type cols_idx: int
        :type record_area: List[int]
        :rtype: None, void
        """
        #========================================#
        # Recursion-based DFS traversal function #
        #========================================#
        if (((rows_idx < 0) or (rows_idx >= len_rows)) or ((cols_idx < 0) or (cols_idx >= len_cols))): #Issue/Boundary-case handling
            return

        if ((grid[rows_idx])[cols_idx] != 1): #Issue/Boundary-case handling
            return


        ####################
        #Whole process/flow
        record_area[0] += 1 #Update/Accumulate

        (grid[rows_idx])[cols_idx] = (-1) #Update/Overwrite

        self.dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx - 1), record_area) #Recursion function call (i.e. left)
        self.dfsTraversal(grid, len_rows, len_cols, (rows_idx - 1), cols_idx, record_area) #Recursion function call (i.e. up)
        self.dfsTraversal(grid, len_rows, len_cols, rows_idx, (cols_idx + 1), record_area) #Recursion function call (i.e. right)
        self.dfsTraversal(grid, len_rows, len_cols, (rows_idx + 1), cols_idx, record_area) #Recursion function call (i.e. down)
